File: coding/src/com/algos/core/li07_stack/revised/L155_MinStack.java
package  com.algos.core.li07_stack.revised;

import java.util.Stack;

public class L155_MinStack {

	public static void main(String[] args) {
		L155_MinStack obj = new L155_MinStack();
		obj.push(-2);
		obj.push(0);
		obj.push(-3);
		System.out.println(obj.getMin());
		obj.pop();
		System.out.println(obj.top());
		System.out.println(obj.getMin());
	}

    // Problem: https://leetcode.com/problems/min-stack/ 
    // Idea: Idea is simple. Use 2 stack. One normal stack to support normal stack operations like push, pop and peek.
    // Use another stack which tracks the min till every element. 
    // While pushing of element, min stack pushes min(curr, peek) which will become min till that element.
	Stack<Integer> stack;
    Stack<Integer> minStack;
    public L155_MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }
    
    public void push(int val) {
        stack.push(val);
        if(minStack.isEmpty()){
            minStack.push(val);
        }else{
            minStack.push(Math.min(val, minStack.peek()));
        }
    }
    
    public void pop() {
        stack.pop();
        minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}

File: dsa/concepts/07_stacks/python/l155_min_stack.py
"""
Problem: https://leetcode.com/problems/min-stack/
Idea:
Use two stacks.
One stack supports normal stack operations (push, pop, top).
The second stack keeps track of the minimum value till each element.
While pushing, store min(current_value, current_min) in the min stack.
"""

class MinStack:

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)

        if not self.min_stack:
            self.min_stack.append(val)
        else:
            self.min_stack.append(min(val, self.min_stack[-1]))

    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def get_min(self) -> int:
        return self.min_stack[-1]


# Example usage
if __name__ == "__main__":
    obj = MinStack()
    obj.push(-2)
    obj.push(0)
    obj.push(-3)
    print(obj.get_min())  # -3
    obj.pop()
    print(obj.top())      # 0
    print(obj.get_min())  # -2
